### PR TITLE
Publir Bid Adapter : initial release

### PR DESCRIPTION
**Pull request by mdghousesaqlain**
## Type of change

- [x] Feature

## Description of change

- contact email : info@publir.com
- test parameters for validating bids:

## How to test
Use Google Chrome
Go to https://www.cravehub.com/test_prebid.html, we are loading Publir Bid Adapter.
```
{
  bidder: 'publir',
  params: {
    pubId: '13301c0d200f243a475d5c5f'
  }
}
```

## Other information

[-> Docs PR ](https://github.com/prebid/prebid.github.io/pull/5043)

### CODE REVIEW

**Comment by mdghousesaqlain** [comment]
@osazos 
Hey, Can we have an status update on the approval of our adapter. 
We are hoping to get the adapter live end of January.

File: integrationExamples/gpt/publir_hello_world.html [hunk 1]
```diff
@@ -0,0 +1,83 @@
+<html>
+<head>
+    <script async src="../../build/dev/prebid.js"></script>
+    <script async src="https://www.googletagservices.com/tag/js/gpt.js"></script>
```
Comment (patmmccann) [inline integrationExamples/gpt/publir_hello_world.html:4]: This is the wrong location of gpt

File: modules/publirBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+const CURRENCY = 'USD';
+const DEFAULT_SELLER_ENDPOINT = 'https://prebid.publir.com/publirPrebidEndPoint';
+const SUPPORTED_SYNC_METHODS = {
+  IFRAME: 'iframe',
+  PIXEL: 'pixel'
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: 32,
```
Comment (patmmccann) [inline modules/publirBidAdapter.js:31]: This is Microsoft. @jsnellbaker have you authorized someone to use your gvlid?

Comment (jsnellbaker) [reply modules/publirBidAdapter.js:31]: Not to my knowledge.

**Review by osazos — COMMENTED** [review]
Hi @mdghousesaqlain,

- I got a CORS error on your endpoint, can you check it?
- please add more tests to increase coverage rate
- regarding your [documentation](https://github.com/prebid/prebid.github.io/pull/5043/files):
  - you should mention "video" in addition of "banner" as supported media_types
  - add `coppa_supported: true`, your code seems to support it

File: modules/publirBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+/**
+   * Generate params that are common between all bids
+   * @param {single bid object} generalObject
+   * @param {bidderRequest} bidderRequest
+   * @returns {object} the common params object
+   */
+function generatePubGeneralsParams(generalObject, bidderRequest) {
+  const domain = window.location.hostname;
```
[the remark this answers is not in the document]
Comment (osazos) [reply modules/publirBidAdapter.js]: Please don't resolve conversation by yourself.

The modification you did is not correct, `refererInfo` is an object which provide a lot of information. Take a look a this.

[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Okay, I'm taking all the info but using only `bidderRequest.refererInfo.domain` on the endpoint, as i need only domain info

File: modules/publirBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        };
+
+        if (adUnit.mediaType === VIDEO) {
+          bidResponse.vastXml = adUnit.vastXml;
+        } else if (adUnit.mediaType === BANNER) {
+          bidResponse.ad = adUnit.ad;
+        }
+
+        if (adUnit.adomain && adUnit.adomain.length) {
+          bidResponse.meta.advertiserDomains = adUnit.adomain;
```
[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Updated this as `bidResponse.meta.advertiserDomains = [adUnit.adomain]`

File: modules/publirBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    if (body.bids) {
+      body.bids.forEach(adUnit => {
+        const bidResponse = {
+          requestId: adUnit.requestId,
+          cpm: adUnit.cpm,
+          currency: adUnit.currency || CURRENCY,
+          width: adUnit.width,
+          height: adUnit.height,
+          ttl: adUnit.ttl || TTL,
+          creativeId: adUnit.requestId,
```
[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: It is not expected, Now I have changed it.

File: modules/publirBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+  const generalParams = {
+    wrapper_type: 'prebidjs',
+    wrapper_vendor: '$$PREBID_GLOBAL$$',
+    wrapper_version: '$prebid.version$',
+    adapter_version: ADAPTER_VERSION,
+    auction_start: timestamp(),
+    publisher_id: generalBidParams.pubId,
+    publisher_name: domain,
+    site_domain: domain,
+    dnt: (navigator.doNotTrack == 'yes' || navigator.doNotTrack == '1' || navigator.msDoNotTrack == '1') ? 1 : 0,
```
[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Thank you, now using getDNT() from utils.

File: modules/publirBidAdapter.js [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+  const timeout = bidderRequest.timeout;
+
+  // these params are snake_case instead of camelCase to allow backwards compatability on the server.
+  // in the future, these will be converted to camelCase to match our convention.
+  const generalParams = {
+    wrapper_type: 'prebidjs',
+    wrapper_vendor: '$$PREBID_GLOBAL$$',
+    wrapper_version: '$prebid.version$',
+    adapter_version: ADAPTER_VERSION,
+    auction_start: timestamp(),
```
[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Thank you, now using the same as suggested.

File: modules/publirBidAdapter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (osazos) [reply modules/publirBidAdapter.js]: I was thinking to add a else case like

```js
if (adUnit.adomain && adUnit.adomain.length) {
          bidResponse.meta.advertiserDomains = adUnit.adomain;
} else {
  bidResponse.meta.advertiserDomains = [];
}
```

File: modules/publirBidAdapter.js [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+      }
+    }
+    return syncs;
+  },
+  onBidWon: function (bid) {
+    if (bid == null) {
+      return;
+    }
+    logInfo('onBidWon:', bid);
+    fetch('//wsfd8lvwt6.execute-api.us-east-1.amazonaws.com/default/publirPrebidImpressionTracker', { method: 'POST', mode: 'no-cors', body: JSON.stringify(bid), credentials: 'include', headers: { 'Content-Type': 'application/json' } });
```
Comment (osazos) [inline modules/publirBidAdapter.js:134]: Please use `ajax()` from `import {ajax} from '../src/ajax.js';`

File: modules/publirBidAdapter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Thank you for the suggestion, I'll update it.

File: modules/publirBidAdapter.js (hunk 8 above)
Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: Please review, code modified as suggested.

**Comment by mdghousesaqlain** [comment]
@ChrisHuie @osazos 
Can you please review the changes and let us know the feedback, if everything good, can you please approve the PR.

**Comment by osazos** [comment]
Hi @mdghousesaqlain sounds good.

But there is sill the point about your coverage rate, it is expected to be > 80%. 

> code under review has at least 80% unit test coverage

Can you improve it? After that, it should be ok.

File: modules/publirBidAdapter.js [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+  getUserSyncs: function (syncOptions, serverResponses) {
+    const syncs = [];
+    for (const response of serverResponses) {
+      if (syncOptions.iframeEnabled && response.body.params.userSyncURL) {
+        syncs.push({
+          type: 'iframe',
+          url: response.body.params.userSyncURL
+        });
+      }
+      if (syncOptions.pixelEnabled && isArray(response.body.params.userSyncPixels)) {
```
Comment (osazos) [inline modules/publirBidAdapter.js:117]: Last one I think… you don't valid body.params exists, it throws an error when your response is empty

Capture d’écran 2024-03-06 à 12 02 54

Comment (mdghousesaqlain) [reply modules/publirBidAdapter.js]: As per the feedback, modified the code.

Comment (osazos) [reply modules/publirBidAdapter.js]: I apologize, but you also have to do it for userSyncURL

**Comment by mdghousesaqlain** [comment]
@osazos 
Please let us know, if any modification / suggestion needs to be done.
We are eagerly waiting for the adapter to be approved.

**Review by osazos — APPROVED** [review]
LGTM

**Pull request merged**